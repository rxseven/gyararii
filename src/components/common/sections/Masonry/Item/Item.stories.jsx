import React from 'react';
import { action } from '@storybook/addon-actions';
import { storiesOf } from '@storybook/react';

import Item from './index';

const title = 'Components/Common/Sections/Masonry/Item';
const props = {
  image: {
    public_id: 'image-001',
    secure_url: 'https://via.placeholder.com/480x320?text=Image-001',
    url: 'https://via.placeholder.com/480x320?text=Image-001'
  },
  index: 1,
  onPreview: action('onPreview'),
  onSelect: action('onSelect')
};

storiesOf(title, module)
  .add('default', () => <Item {...props} />)
  .add('selected', () => <Item {...props} isSelected />)
  .add('to delete', () => <Item {...props} isSelected toDelete />)
  .add('deleting', () => <Item {...props} isLoading isSelected toDelete />);
