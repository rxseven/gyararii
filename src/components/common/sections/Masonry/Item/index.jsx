/* eslint-disable no-nested-ternary */
import PropTypes from 'prop-types';
import exact from 'prop-types-exact';
import React, { useState } from 'react';
import styled, { css } from 'styled-components';
import { isMobile } from 'react-device-detect';

import Card from 'components/common/base/Card';
import Icon from 'components/common/base/Icon';
import Spinner from 'components/common/base/Spinner';

const styles = {
  icon: css`
    color: #fff;
    cursor: pointer;
    display: ${props => (props['data-visibility'] ? 'block' : 'none')};
    font-size: ${props => (props['data-mobile'] ? '1.5rem' : '1rem')};
    position: absolute;
  `
};

const Frame = styled.div`
  border-bottom-left-radius: ${({ theme }) => theme.border.radius};
  border-bottom-right-radius: ${({ theme }) => theme.border.radius};
  cursor: pointer;
  position: relative;
  transition: box-shadow 0.15s linear;
  transition-delay: 0.05s;

  :hover {
    box-shadow: 0 4px 6px rgba(0, 0, 0, 0.3);
  }
`;

const Link = styled.a`
  ${styles.icon}
  color: ${props => (props['data-mobile'] ? '#666' : '#fff')};
  display: none;
  left: ${props => (props['data-mobile'] ? '42px' : '32px')};
  opacity: ${props => (props['data-mobile'] ? 0.5 : 1)};
  top: 2px;

  /* stylelint-disable-next-line */
  ${Frame}:hover & {
    display: ${props => (props['data-mobile'] ? 'none' : 'block')};
    opacity: 0.5;

    :hover {
      color: #fff;
      opacity: 1;
    }
  }
`;

const Overlay = styled.div`
  align-items: center;
  background-color: ${({ isLoading }) =>
    isLoading ? 'rgba(0,0,0,0.3)' : 'transparent'};
  bottom: 0;
  display: ${({ isLoading }) => (isLoading ? 'flex' : 'none')};
  justify-content: center;
  left: 0;
  position: absolute;
  right: 0;
  top: 0;
`;

const Select = styled(Icon)`
  ${styles.icon}
  color: ${props =>
    props['data-checked'] ? '#ff620c' : props['data-mobile'] ? '#666' : '#fff'};
  display: ${props =>
    props['data-checked'] || props['data-mobile'] ? 'block' : 'none'};
  left: 7px;
  opacity: ${props => (props['data-mobile'] ? 0.5 : 1)};
  top: 7px;

  /* stylelint-disable-next-line */
  ${Frame}:hover & {
    display: block;
    opacity: ${props => (props['data-checked'] ? '1' : '0.5')};

    :hover {
      color: ${props => (props['data-checked'] ? '#ff620c' : '#fff')};
      opacity: 1;
    }
  }
`;

const Shadow = styled.div`
  box-shadow: inset 0px 85px 95px -50px rgba(0, 0, 0, 0.65);
  border-top-left-radius: ${({ theme }) => theme.border.radius};
  border-top-right-radius: ${({ theme }) => theme.border.radius};
  display: none;
  height: 100%;
  left: 0;
  position: absolute;
  top: 0;
  width: 100%;
  z-index: 0;

  /* stylelint-disable-next-line */
  ${Frame}:hover & {
    display: ${({ mobile }) => (mobile ? 'none' : 'block')};
  }
`;

const Thumbnail = styled(Card)`
  &&& {
    border: ${props =>
      props['data-visibility'] ? '1px solid rgba(0, 0, 0, 0.125)' : 'none'};
    position: ${props => (props['data-visibility'] ? 'static' : 'absolute')};
    visibility: ${props => (props['data-visibility'] ? 'visible' : 'hidden')};
    z-index: -999;
  }
`;

const propTypes = exact({
  image: PropTypes.shape({
    public_id: PropTypes.string.isRequired,
    secure_url: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired
  }).isRequired,
  index: PropTypes.number.isRequired,
  isLoading: PropTypes.bool,
  onLoaded: PropTypes.func.isRequired,
  onPreview: PropTypes.func.isRequired,
  isSelected: PropTypes.bool,
  onSelect: PropTypes.func.isRequired,
  toDelete: PropTypes.bool
});

const defaultProps = {
  isLoading: false,
  isSelected: false,
  toDelete: false
};

function Item(props) {
  const [visibility, setVisible] = useState(false);
  const {
    image: { public_id: publicId, secure_url: https, url },
    index,
    onLoaded,
    isLoading,
    onPreview,
    onSelect,
    isSelected,
    toDelete
  } = props;

  function handleLoaded() {
    onLoaded();
    setVisible(true);
  }

  function handleSelect(event) {
    event.stopPropagation();
    onSelect(publicId);
  }

  function handleOpenLink(event) {
    event.stopPropagation();
  }

  function handleOpenLightbox() {
    onPreview(index);
  }

  return (
    <Thumbnail data-visibility={visibility}>
      <Frame data-testid="frame" onClick={handleOpenLightbox}>
        <Card.Image src={https} onLoad={handleLoaded} />
        <Shadow mobile={isMobile} />
        <Select
          data-checked={isSelected}
          data-mobile={isMobile}
          data-testid="select"
          icon={toDelete ? 'trash-alt' : 'check-circle'}
          onClick={handleSelect}
        />
        <Link
          data-mobile={isMobile}
          href={url}
          onClick={handleOpenLink}
          target="_blank"
          title="Open image in new tab"
        >
          <Icon icon="external-link-alt" />
        </Link>
        <Overlay data-testid="overlay" isLoading={isLoading && isSelected}>
          <Spinner color="#fff" />
        </Overlay>
      </Frame>
    </Thumbnail>
  );
}

Item.propTypes = propTypes;
Item.defaultProps = defaultProps;

export default Item;
