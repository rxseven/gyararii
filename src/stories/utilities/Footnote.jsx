import styled from 'styled-components';

const Footnote = styled.div`
  bottom: 2rem;
  font-size: 14px;
  position: fixed;
`;

export default Footnote;
