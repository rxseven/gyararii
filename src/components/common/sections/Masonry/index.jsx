import PropTypes from 'prop-types';
import React from 'react';
import styled from 'styled-components';

import Card from 'components/common/base/Card';
import Item from './Item';

const Frame = styled.div`
  margin-bottom: 0;
`;

const Label = styled.p`
  color: #666;
  font-size: 0.875rem;
  text-align: ${({ center }) => (center ? 'center' : 'left')};
`;

const Message = styled.div`
  margin-top: 1.75rem;
`;

const propTypes = {
  error: PropTypes.number,
  images: PropTypes.arrayOf(PropTypes.object).isRequired,
  isDeleting: PropTypes.bool,
  isFetching: PropTypes.bool,
  isMore: PropTypes.bool,
  selected: PropTypes.arrayOf(PropTypes.string)
};

const defaultProps = {
  error: 0,
  isDeleting: false,
  isFetching: false,
  isMore: false,
  selected: []
};

function Masonry({
  error,
  images,
  isDeleting,
  isFetching,
  isMore,
  selected,
  ...props
}) {
  const imageLength = images.length;
  const isAny = imageLength > 0;
  const isLoaded = error === 0 && !isFetching;
  const isEmpty = !isAny && isLoaded && !isMore;
  const isRemaining = !isAny && isMore && isLoaded;

  return (
    <Frame>
      <If condition={isRemaining}>
        <Message>
          <Label center>Remaining images will be loaded automatically.</Label>
        </Message>
      </If>
      <If condition={isEmpty}>
        <Label>No images found</Label>
      </If>
      <Card.Columns>
        <Choose>
          <When condition={imageLength > 0}>
            {images.map((image, index) => {
              const isSelected = selected.includes(image.public_id);

              return (
                <Item
                  {...props}
                  image={image}
                  index={index}
                  isLoading={isDeleting}
                  isSelected={isSelected}
                  key={image.public_id}
                />
              );
            })}
          </When>
          <Otherwise>
            <div />
          </Otherwise>
        </Choose>
      </Card.Columns>
    </Frame>
  );
}

Masonry.propTypes = propTypes;
Masonry.defaultProps = defaultProps;

export default Masonry;
