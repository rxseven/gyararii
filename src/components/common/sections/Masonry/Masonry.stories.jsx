import React from 'react';
import { action } from '@storybook/addon-actions';
import { storiesOf } from '@storybook/react';

import Masonry from './index';

const title = 'Components/Common/Sections/Masonry';
const props = {
  images: [
    {
      public_id: 'image-001',
      secure_url: 'https://via.placeholder.com/480x320?text=Image-001',
      url: 'https://via.placeholder.com/480x320?text=Image-001'
    },
    {
      public_id: 'image-002',
      secure_url: 'https://via.placeholder.com/480x320?text=Image-002',
      url: 'https://via.placeholder.com/480x320?text=Image-002'
    }
  ],
  onPreview: action('onPreview'),
  onSelect: action('onSelect')
};

storiesOf(title, module)
  .add('has images', () => <Masonry {...props} />)
  .add('empty', () => <Masonry {...props} images={[]} />)
  .add('remaining images', () => <Masonry {...props} images={[]} isMore />);
