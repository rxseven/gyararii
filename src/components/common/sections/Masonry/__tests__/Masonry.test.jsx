import React from 'react';
import { factory } from 'tests/utilities';
import Item from '../Item';
import Masonry from '../index';

// Mock
jest.mock('../Item', () => jest.fn(() => <i />));

// Arrange
const source = {
  images: [
    {
      public_id: 'image-001',
      secure_url: 'gallery/image-001.jpg',
      url: 'gallery/image-001.jpg'
    }
  ],
  onLoaded: jest.fn(),
  onPreview: jest.fn(),
  onSelect: jest.fn()
};
const input = { ...source };

// Setup
function setup(props) {
  return factory(Masonry, source, props);
}

// Cleanup
afterEach(() => {
  Item.mockClear();
});

// Test suites
describe('<Masonry />', () => {
  it('should render without crashing', () => {
    setup();
  });

  it('should render one image', () => {
    const expected = { called: input.images.length };
    setup();

    expect(Item).toHaveBeenCalledTimes(expected.called);
  });

  it('should render two images', () => {
    const props = {
      images: [
        ...source.images,
        {
          public_id: 'image-002',
          secure_url: 'gallery/image-002.jpg',
          url: 'gallery/image-002.jpg'
        }
      ]
    };
    const expected = { called: props.images.length };
    setup(props);

    expect(Item).toHaveBeenCalledTimes(expected.called);
  });

  it('should render empty container when there are no images', () => {
    const props = { images: [] };
    const { queryByTestId } = setup(props);

    expect(queryByTestId('empty')).not.toBeNull();
  });

  it('should render a message when there are no images', () => {
    const props = { images: [] };
    const expected = { content: 'No images found' };
    const { component } = setup(props);

    expect(component).toHaveTextContent(expected.content);
  });

  it('should render a message when there are some images to be loaded automatically', () => {
    const props = {
      images: [],
      isMore: true
    };
    const expected = {
      content: 'Remaining images will be loaded automatically.'
    };
    const { component } = setup(props);

    expect(component).toHaveTextContent(expected.content);
  });
});
