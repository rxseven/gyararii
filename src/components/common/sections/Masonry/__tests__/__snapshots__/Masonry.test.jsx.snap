// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Masonry /> Snapshot tests should render correctly 1`] = `
<div
  className="sc-bZQynM ghYVpU"
>
  <div
    className="card-columns"
  >
    <div
      className="card sc-EHOje bJcwWL text-text-left mb-3"
    >
      <div
        className="sc-bdVaJa cLhofW"
        onClick={[Function]}
      >
        <div
          className="sc-ifAKCX dNDTYw"
        />
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-check-circle fa-w-16 sc-bxivhb gbCqig"
          data-checked={false}
          data-icon="check-circle"
          data-mobile={false}
          data-prefix="fas"
          focusable="false"
          onClick={[Function]}
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
        <a
          className="sc-bwzfXH fbJkSt"
          data-mobile={false}
          href="gallery/image-001.jpg"
          onClick={[Function]}
          target="_blank"
          title="Open image in new tab"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-external-link-alt fa-w-18 "
            data-icon="external-link-alt"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 576 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M576 24v127.984c0 21.461-25.96 31.98-40.971 16.971l-35.707-35.709-243.523 243.523c-9.373 9.373-24.568 9.373-33.941 0l-22.627-22.627c-9.373-9.373-9.373-24.569 0-33.941L442.756 76.676l-35.703-35.705C391.982 25.9 402.656 0 424.024 0H552c13.255 0 24 10.745 24 24zM407.029 270.794l-16 16A23.999 23.999 0 0 0 384 303.765V448H64V128h264a24.003 24.003 0 0 0 16.97-7.029l16-16C376.089 89.851 365.381 64 344 64H48C21.49 64 0 85.49 0 112v352c0 26.51 21.49 48 48 48h352c26.51 0 48-21.49 48-48V287.764c0-21.382-25.852-32.09-40.971-16.97z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </a>
        <div
          className="sc-htpNat ddGrH"
        >
          <span>
            <style>
               @-webkit-keyframes brsk-three-bounce { 0%, 80%, 100% { -webkit-transform: scale(0); } 40% { -webkit-transform: scale(1); } } @keyframes brsk-three-bounce { 0%, 80%, 100% { transform: scale(0); -webkit-transform: scale(0); } 40% { transform: scale(1); -webkit-transform: scale(1); } } 
            </style>
            <div
              style={
                Object {
                  "WebkitAnimationDelay": "-0.32s",
                  "WebkitAnimationDuration": "1.25s",
                  "WebkitAnimationIterationCount": "infinite",
                  "WebkitAnimationName": "brsk-three-bounce",
                  "WebkitAnimationTimingFunction": "ease-in-out",
                  "animationDelay": "-0.32s",
                  "animationDuration": "1.25s",
                  "animationIterationCount": "infinite",
                  "animationName": "brsk-three-bounce",
                  "animationTimingFunction": "ease-in-out",
                  "backgroundColor": "#fff",
                  "borderRadius": "100%",
                  "display": "inline-block",
                  "height": 7,
                  "marginRight": 5,
                  "width": 7,
                }
              }
            />
            <div
              style={
                Object {
                  "WebkitAnimationDelay": "-0.16s",
                  "WebkitAnimationDuration": "1.25s",
                  "WebkitAnimationIterationCount": "infinite",
                  "WebkitAnimationName": "brsk-three-bounce",
                  "WebkitAnimationTimingFunction": "ease-in-out",
                  "animationDelay": "-0.16s",
                  "animationDuration": "1.25s",
                  "animationIterationCount": "infinite",
                  "animationName": "brsk-three-bounce",
                  "animationTimingFunction": "ease-in-out",
                  "backgroundColor": "#fff",
                  "borderRadius": "100%",
                  "display": "inline-block",
                  "height": 7,
                  "marginRight": 5,
                  "width": 7,
                }
              }
            />
            <div
              style={
                Object {
                  "WebkitAnimationDuration": "1.25s",
                  "WebkitAnimationIterationCount": "infinite",
                  "WebkitAnimationName": "brsk-three-bounce",
                  "WebkitAnimationTimingFunction": "ease-in-out",
                  "animationDuration": "1.25s",
                  "animationIterationCount": "infinite",
                  "animationName": "brsk-three-bounce",
                  "animationTimingFunction": "ease-in-out",
                  "backgroundColor": "#fff",
                  "borderRadius": "100%",
                  "display": "inline-block",
                  "height": 7,
                  "width": 7,
                }
              }
            />
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
